### PR TITLE
Advanced search queries

### PR DESCRIPTION
**Pull request by juliushaertl**
Based on #2933 

- Implement advanced search queries (see detailed description below)
- Proper pagination in unified search providers
- Implement card comment search in unified search
- Add global search result UI to deck
- Extend card details route in frontend to navigate to a specific sidebar tab directly
- Some refactoring on the integration tests in order to writing the search tests and further ones more straight forward

## Global search in Deck

image

A dedicated inline search input was added in order to properly indicate that a search filtering is still active even if the unified search is closed:

image

## Search queries

The search will only return results that match all of the search tokens.

Example query: `project tag:ToDo assigned:alice assigned:bob`

Will return all cards where the card title or description contains project AND the tag ToDo is set AND the user alice is assigned AND the user bob is assigned

| Filter      | Operators         | Query                                                        | Tests |
| ----------- | ----------------- | ------------------------------------------------------------ | ----- |
| title       | `:`               | text token used for a case-insentitive search on the cards title | x     |
| description | `:`               | text token used for a case-insentitive search on the cards description | x     |
| list        | `:`               | text token used for a case-insentitive search on the cards list name | x     |
| tag         | `:`               | text token used for a case-insentitive search on the assigned tags | x     |
| date        | `:`               | 'overdue', 'today', 'week', 'month', 'none'                  | x     |
|             | `>` `<` `>=` `<=` | Compare the card due date to the passed date (see [supported date formats](https://www.php.net/manual/de/datetime.formats.php)) Card due dates are always considered UTC for comparison | x     |
| assigned    | `:`               | id or displayname of a user or group for a search on the assigned users or groups | x     |

Other text tokens will be used to perform a case-insensitive search on the card title and description

Quotes can be used to pass a query with spaces, e.g. `title:"My card"`

### CODE REVIEW

File: lib/Controller/SearchController.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	public function __construct(string $appName, IRequest $request, SearchService $searchService) {
+		parent::__construct($appName, $request);
+		$this->searchService = $searchService;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 */
+	public function search(string $term, int $limit = null, int $cursor = null): DataResponse {
```
Comment (nickvergessen) [inline lib/Controller/SearchController.php:50]:
```suggestion
	public function search(string $term, ?int $limit = null, ?int $cursor = null): DataResponse {
```

File: lib/Search/DeckProvider.php [hunk 2]
```diff
@@ -77,37 +63,34 @@ public function getName(): string {
...
+			}, $boardResults),
+			array_map(function (Card $card) {
+				return new CardSearchResultEntry($card->getRelatedBoard(), $card->getRelatedStack(), $card, $this->urlGenerator);
+			}, $cardResults)
 		);
 
-		return SearchResult::complete(
+		if (count($cardResults) < $query->getLimit()) {
+			return SearchResult::complete(
+				'Deck',
```
Comment (nickvergessen) [inline lib/Search/DeckProvider.php:80]: Translate?

File: lib/Search/DeckProvider.php [hunk 3]
```diff
@@ -77,37 +63,34 @@ public function getName(): string {
...
+				$results
+			);
+		}
+		
+		return SearchResult::paginated(
 			'Deck',
```
Comment (nickvergessen) [inline lib/Search/DeckProvider.php:86]: same here

**Review by PVince81 — COMMENTED** [review]
partial review

File: src/store/card.js [hunk 4]
```diff
@@ -86,8 +87,90 @@ export default {
...
+						} else if (filter === 'list') {
+							const stack = this.getters.stackById(card.stackId)
+							if (!stack) {
+								return false
+							}
+							hasMatch = hasMatch && stack.title.toLowerCase().includes(filterOutQuotes(query).toLowerCase())
+						} else if (filter === 'tag') {
+							hasMatch = hasMatch && card.labels.findIndex((label) => label.title.toLowerCase().includes(filterOutQuotes(query).toLowerCase())) !== -1
+						} else if (filter === 'date') {
+							const datediffHour = ((new Date(card.duedate) - new Date()) / 3600 / 1000)
```
Comment (PVince81) [inline src/store/card.js:115]: maybe use moment for diffing dates to avoid potential troubles when crossing DST offsets (rare, so probably not cricial)

File: src/store/card.js [hunk 5]
```diff
@@ -86,8 +87,90 @@ export default {
...
+							if (card.duedate === null || !hasMatch) {
+								return false
+							}
+							const comparator = query[0] + (query[1] === '=' ? '=' : '')
+							const isValidComparator = ['<', '<=', '>', '>='].indexOf(comparator) !== -1
+							const parsedCardDate = moment(card.duedate)
+							const parsedDate = moment(query.substr(isValidComparator ? comparator.length : 0))
+							switch (comparator) {
+							case '<':
+								hasMatch = hasMatch && parsedCardDate.isBefore(parsedDate)
```
Comment (PVince81) [inline src/store/card.js:144]: :+1: I like. a bit more readable than the logic on the server side

File: src/store/card.js [hunk 6]
```diff
@@ -86,8 +87,90 @@ export default {
...
+							case 'none':
+								hasMatch = hasMatch && (card.duedate === null)
+								break
+							}
+
+							if (card.duedate === null || !hasMatch) {
+								return false
+							}
+							const comparator = query[0] + (query[1] === '=' ? '=' : '')
+							const isValidComparator = ['<', '<=', '>', '>='].indexOf(comparator) !== -1
```
Comment (PVince81) [inline src/store/card.js:139]: throw error in case of no valid comparator ?

File: tests/integration/features/search.feature [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    And assign the tag "Finished" to the card
+    And assign the tag "To review" to the card
+    When searching for 'tag:Finished tag:Later'
+    Then the card "Example task 1" is not found
+    And the card "Multi labeled card" is not found
+
+    When searching for 'tag:Finished tag:"To review"'
+    Then the card "Example task 1" is not found
+    And the card "Labeled card" is not found
+    And the card "Multi labeled card" is found
```
Comment (PVince81) [inline tests/integration/features/search.feature:259]: also add tests for invalid expressions ? (if not covered by unit tests)

**Review by PVince81 — COMMENTED** [review]
second part of review

File: lib/Search/DeckProvider.php (hunk 2 above)
Comment (juliushaertl) [reply lib/Search/DeckProvider.php:80]: I guess that is a different discussion, as we currently never translate the app name for deck.

File: lib/Db/CardMapper.php [hunk 8]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+
+	public function searchComments(array $boardIds, SearchQuery $query, int $limit = null, int $offset = null): array {
+		if (count($query->getTextTokens()) === 0) {
+			return [];
+		}
+		$qb = $this->queryCardsByBoards($boardIds);
+		$this->extendQueryByFilter($qb, $query);
+
+		$qb->innerJoin('c', 'comments', 'comments', $qb->expr()->andX(
+			$qb->expr()->eq('comments.object_id', $qb->createFunction('CAST(c.id as CHAR(50))')),
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Db/CardMapper.php:323]: Should be according to https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/CAST.html#GUID-5A70235E-1209-4281-8521-B94497AAEF75 but I'll create a follow up issue to actually run tests against oracle and give it some test run

File: lib/Db/CardMapper.php [hunk 9]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+
+		foreach ($query->getStack() as $stack) {
+			$qb->andWhere($qb->expr()->iLike('s.title', $qb->createNamedParameter('%' . $this->db->escapeLikeParameter($stack->getValue()) . '%', IQueryBuilder::PARAM_STR)));
+		}
+
+		if (count($query->getTag())) {
+			foreach ($query->getTag() as $index => $tag) {
+				$qb->innerJoin('c', 'deck_assigned_labels', 'al' . $index, $qb->expr()->eq('c.id', 'al' . $index . '.card_id'));
+				$qb->innerJoin('al'. $index, 'deck_labels', 'l' . $index, $qb->expr()->eq('al' . $index . '.label_id', 'l' . $index . '.id'));
+				$qb->andWhere($qb->expr()->iLike('l' . $index . '.title', $qb->createNamedParameter('%' . $this->db->escapeLikeParameter($tag->getValue()) . '%', IQueryBuilder::PARAM_STR)));
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Db/CardMapper.php:378]: The array will always use an implicit integer key https://github.com/nextcloud/deck/blob/e111e22077d21633025e035a60fd52ab8e6a9289/lib/Search/Query/SearchQuery.php#L86-L92

File: lib/Db/CardMapper.php [hunk 10]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+
+		if (count($query->getAssigned()) > 0) {
+			foreach ($query->getAssigned() as $index => $assignment) {
+				$qb->innerJoin('c', 'deck_assigned_users', 'au' . $index, $qb->expr()->eq('c.id', 'au' . $index . '.card_id'));
+				$assignedQueryValue = $assignment->getValue();
+				$searchUsers = $this->userManager->searchDisplayName($assignment->getValue());
+				$users = array_filter($searchUsers, function (IUser $user) use ($assignedQueryValue) {
+					return (mb_strtolower($user->getDisplayName()) === mb_strtolower($assignedQueryValue) || $user->getUID() === $assignedQueryValue);
+				});
+				$groups = $this->groupManager->search($assignment->getValue());
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Db/CardMapper.php:429]: Yep, this depends on the user backend implementation but the default database and ldap group backend both search for the gid and displayname inside of the managers search method.

File: src/components/search/Placeholder.vue [hunk 11]
```diff
@@ -0,0 +1,115 @@
+<template>
+	<div class="card--placeholder">
+		<svg class="card-placeholder__gradient">
+			<defs>
+				<linearGradient id="card-placeholder__gradient">
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply src/components/search/Placeholder.vue:5]: Will do a follow up ticket for this and the unified search in the server where this was inspired from

File: lib/Db/CardMapper.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/Db/CardMapper.php:323]: https://github.com/nextcloud/server/blob/master/lib/private/DB/QueryBuilder/ExpressionBuilder/OCIExpressionBuilder.php#L41

Just specify the type on the eq() method and it should work automatically

File: lib/Db/CardMapper.php [hunk 12]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+		}
+
+		foreach ($query->getDuedate() as $duedate) {
+			$date = $duedate->getValue();
+			$supportedFilters = ['overdue', 'today', 'week', 'month', 'none'];
+			if (in_array($date, $supportedFilters, true)) {
+				if ($date === 'overdue') {
+					$qb->andWhere($qb->expr()->lt('c.duedate', $qb->createNamedParameter((new \DateTime()), IQueryBuilder::PARAM_DATE)));
+				} elseif ($date === 'today') {
+					$qb->andWhere($qb->expr()->gte('c.duedate', $qb->createNamedParameter((new \DateTime()), IQueryBuilder::PARAM_DATE)));
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/Db/CardMapper.php:389]: Also tehre is ITimeFactory::getDatetime()

File: lib/Db/CardMapper.php [hunk 13]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+			);
+		}
+
+		$qb->groupBy('c.id');
+		$qb->orderBy('c.last_modified', 'DESC');
+		if ($limit !== null) {
+			$qb->setMaxResults($limit);
+		}
+		if ($offset !== null) {
+			$qb->andWhere($qb->expr()->lt('c.last_modified', $qb->createNamedParameter($offset, IQueryBuilder::PARAM_INT)));
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Db/CardMapper.php:303]: Filed https://github.com/nextcloud/deck/issues/2968 for this

File: tests/integration/features/search.feature (hunk 7 above)
Comment (juliushaertl) [reply tests/integration/features/search.feature:259]: There are not really invalid expressions since everything that is not being parsed is added as a simple text token for searching in the plain text fields of the card, but I'll cover the parser with some additional unit tests.

File: src/components/search/GlobalSearchResults.vue [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+	- GNU Affero General Public License for more details.
+	-
+	- You should have received a copy of the GNU Affero General Public License
+	- along with this program. If not, see <http://www.gnu.org/licenses/>.
+	-
+	-->
+
+<template>
+	<div v-if="searchQuery!==''" class="global-search">
+		<h2><RichText :text="t('deck', 'Search for {searchQuery} in all boards')" :arguments="queryStringArgs" /></h2>
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply src/components/search/GlobalSearchResults.vue:25]: Ah missing the styles are not applying as expected, will push a highlighting like this as quotes could look odd once the users uses quotes in the string itself:

image

File: tests/integration/features/search.feature (hunk 7 above)
Comment (juliushaertl) [reply tests/integration/features/search.feature:259]: Added some testing of to the parser be2f070

File: lib/Service/SearchService.php [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+		$matchedComments = $this->cardMapper->searchComments($boardIds, $this->filterStringParser->parse($term), $limit, $cursor);
+
+		$self = $this;
+		return array_map(function ($cardRow) use ($self) {
+			$comment = $this->commentsManager->get($cardRow['comment_id']);
+			unset($cardRow['comment_id']);
+			$card = Card::fromRow($cardRow);
+			$self->cardService->enrich($card);
+			$user = $this->userManager->get($comment->getActorId());
+			$displayName = $user ? $user->getDisplayName() : '';
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Service/SearchService.php:116]: For reference https://github.com/nextcloud/server/issues/10021

Will add to the follow up list.

File: lib/Search/FilterStringParser.php [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+	public function __construct(IL10N $l10n) {
+		$this->l10n = $l10n;
+	}
+	
+	public function parse(?string $filter): SearchQuery {
+		$query = new SearchQuery();
+		if (empty($filter)) {
+			return $query;
+		}
+		$tokens = preg_split('/\s(?=([^"]*"[^"]*")*[^"]*$)/', $filter);
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Search/FilterStringParser.php]: I've pushed some refinement of the regex to handle quotes a bit more gracefully as i found when writing the tests the old one had some edge cases with unclosed quotes breaking the matches.

File: src/store/card.js (hunk 6 above)
Comment (juliushaertl) [reply src/store/card.js:139]: Until now invalid queries would just attempt to perform an equal comparison for whatever is passed after the `:`. I'd keep that for now and move the handling improvement of that into a ticket so we can discuss where and how to show possible errors especially with the unified search, where we'd need server improvements to return possible hints or errors to the user.

File: lib/Search/FilterStringParser.php [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+		$type = strtolower($type);
+
+		$qualifier = null;
+
+		switch ($type) {
+			case 'date':
+				$comparator = SearchQuery::COMPARATOR_EQUAL;
+				$value = $param;
+				if ($param[0] === '<' || $param[0] === '>') {
+					$orEquals = $param[1] === '=';
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Search/FilterStringParser.php:88]: Moving to a follow up ticket as similar to in https://github.com/nextcloud/deck/pull/2934#discussion_r612204889

File: src/store/card.js (hunk 4 above)
Comment (juliushaertl) [reply src/store/card.js:115]: This is for now using the same logic as the local board filters but I'll split this into a separate ticket for refining both both them.

File: lib/Db/CardMapper.php [hunk 18]
```diff
@@ -260,24 +268,195 @@ public function findUnexposedDescriptionChances() {
...
+					$qb->andWhere($qb->expr()->isNull('c.duedate'));
+				}
+			}
+
+			try {
+				$date = new \DateTime($date);
+				if ($duedate->getComparator() === SearchQuery::COMPARATOR_LESS) {
+					$qb->andWhere($qb->expr()->lt('c.duedate', $qb->createNamedParameter($date, IQueryBuilder::PARAM_DATE)));
+				} elseif ($duedate->getComparator() === SearchQuery::COMPARATOR_LESS_EQUAL) {
+					$date->setTime(23, 59, 59);
```
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Db/CardMapper.php:413]: Added some comment :+1:

**Review by PVince81 — APPROVED** [review]
Code looks good now, thanks for the adjustments :+1:

File: src/store/card.js (hunk 6 above)
Comment (juliushaertl) [reply src/store/card.js:139]: https://github.com/nextcloud/deck/issues/2973

File: src/store/card.js (hunk 4 above)
Comment (juliushaertl) [reply src/store/card.js:115]: https://github.com/nextcloud/deck/issues/2974

File: src/components/search/Placeholder.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (juliushaertl) [reply src/components/search/Placeholder.vue:5]: - Unified search in server https://github.com/nextcloud/server/issues/26545
- Deck https://github.com/nextcloud/deck/issues/2975

File: lib/Service/SearchService.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (juliushaertl) [reply lib/Service/SearchService.php:113]: https://github.com/nextcloud/deck/issues/2977
